lib/auth/userpreferences/userpreferencesv1/service.go

```diff
@@ -21,42 +21,69 @@ import (
 	"context"
 
 	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
 	"google.golang.org/protobuf/types/known/emptypb"
 
 	userpreferences "github.com/gravitational/teleport/api/gen/proto/go/userpreferences/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/authz"
 	"github.com/gravitational/teleport/lib/services"
 )
 
 // ServiceConfig holds configuration options for the user preferences service.
 type ServiceConfig struct {
-	Backend services.UserPreferences
+	Backend    services.UserPreferencesBackend
+	Authorizer authz.Authorizer
+	Logger     *logrus.Entry
 }
 
 // Service implements the teleport.userpreferences.v1.UserPreferencesService RPC service.
 type Service struct {
 	userpreferences.UnimplementedUserPreferencesServiceServer
 
-	backend services.UserPreferences
+	backend    services.UserPreferencesBackend
+	authorizer authz.Authorizer
+	log        *logrus.Entry
 }
 
 // NewService returns a new user preferences gRPC service.
 func NewService(cfg *ServiceConfig) (*Service, error) {
 	switch {
 	case cfg.Backend == nil:
 		return nil, trace.BadParameter("backend is required")
+	case cfg.Authorizer == nil:
+		return nil, trace.BadParameter("authorizer is required")
+	case cfg.Logger == nil:
+		cfg.Logger = logrus.WithField(trace.Component, "userpreferences.service")
 	}
 
 	return &Service{
-		backend: cfg.Backend,
+		backend:    cfg.Backend,
+		authorizer: cfg.Authorizer,
+		log:        cfg.Logger,
 	}, nil
 }
 
 // GetUserPreferences returns the user preferences for a given user.
 func (a *Service) GetUserPreferences(ctx context.Context, req *userpreferences.GetUserPreferencesRequest) (*userpreferences.GetUserPreferencesResponse, error) {
-	return a.backend.GetUserPreferences(ctx, req)
+	authCtx, err := authz.AuthorizeWithVerbs(ctx, a.log, a.authorizer, true, types.KindUser, types.VerbUpdate)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	username := authCtx.User.GetName()
+
+	return a.backend.GetUserPreferences(ctx, username, req)
 }
 
 // UpsertUserPreferences creates or updates user preferences for a given username.
 func (a *Service) UpsertUserPreferences(ctx context.Context, req *userpreferences.UpsertUserPreferencesRequest) (*emptypb.Empty, error) {
-	return &emptypb.Empty{}, trace.Wrap(a.backend.UpsertUserPreferences(ctx, req))
+	authCtx, err := authz.AuthorizeWithVerbs(ctx, a.log, a.authorizer, true, types.KindUser, types.VerbUpdate)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```

Comment:
We should allow users to modify their own user preferences without having `user` `update` privileges.
```suggestion
	authCtx, err := a.authorizer.Authorize(ctx)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	if !authz.IsLocalUser(authCtx) {
		return trace.IsAccessDenied("Non-local user cannot get user preferences")
	}
```

Reply:
Changed